src/edg.rs

```diff
@@ -211,135 +211,167 @@ impl<B: Broker<V> + Debug, G: ExtendedDependencyGraph<V> + Send + Sync + Debug,
         // Alg 1, Line 2
         if self.is_owner(&self.v0.clone()) {
             trace!(worker_id = self.id, "exploring v0");
-            let init_weight = Weight::new_full();
-            self.explore(&self.v0.clone(), init_weight);
+            self.explore(&self.v0.clone());
         }
 
-        let term_rx = self.term_rx.clone();
-        let hyper_rx = self.hyper_rx.clone();
-        let negation_rx = self.negation_rx.clone();
-        let msg_rx = self.msg_rx.clone();
-
         loop {
-            // Termination signal indicating result have been found
-            if !term_rx.is_empty() {
-                self.counter = 0;
-                match term_rx.recv() {
-                    Ok(assignment) => {
-                        // Alg 1, Line 11-12
-                        trace!(?assignment, "worker received termination");
-                        return match assignment {
-                            VertexAssignment::UNDECIDED => VertexAssignment::FALSE,
-                            VertexAssignment::FALSE => VertexAssignment::FALSE,
-                            VertexAssignment::TRUE => VertexAssignment::TRUE,
-                        };
+            // Pump all messages from broker to queues
+            let mut token: Option<MsgToken> = None;
+            loop {
+                match self.msg_rx.try_recv() {
+                    Ok(msg) => {
+                        match msg {
+                            Message::REQUEST { .. } => self.msg_queue.push_back(msg),
+                            Message::ANSWER { .. } => self.msg_queue.push_back(msg),
+                            Message::TOKEN(msg_token) => {
+                                debug_assert!(token.is_none(), "Received multiple tokens. Something is wrong with the token ring");
+                                token = Some(msg_token)
+                            }
+                            Message::RELEASE(_) => self.msg_queue.push_back(msg),
+                            Message::NEGATION(edge) => {
+                                match self.assignment.get(&edge.target) {
+                                    None => self.queue_negation(edge),
+                                    Some(_) => {
+                                        // Edge have assignment UNDECIDED, TRUE, or FALSE, and is there for safe to explore
+                                        self.negation_queue.push_back(edge)
+                                    }
+                                }
+                            }
```

Comment:
There might be an error here. Could you explain how a `Message::NEGATION` ends up in this rx?

Reply:
Interesting, that would happen if `Broker::queue_negation` is called which it never is

Reply:
I am more confused now, thanks.

Reply:
Unsafe negation edges are negation edges, where the target vertex is assigned `undecided`. The `queue_negation()` function puts a negation in the unsafe queue (sidenote: we should add unsafe to the method name). Everything put in the `negation_queue` is processed, even if its target is assigned `undecided` and the edge is unsafe. This is okay, as long as we only queue unsafe on release of unsafe edges.

This is not what is happening here. So, maybe error?

In `release_negations` we are putting the released edges straight into the `negation_queue` so that works fine.

Reply:
... Why do we even have `Message::Negation`? The only thing sent between workers should be requests, answers, token, and terminate.

Reply:
`Message::Negation` and `Message::Hyper` is now deleted. The `Message` enum now correctly describes messages sent between workers.